### PR TITLE
Add Swift Package Manager as new opt-in feature for iOS and macOS

### LINKED ISSUE #146369: Add Swift Package Manager as opt-in feature
Add tooling to allow plugins to be written as Swift Packages and use Swift Package Manager as the dependency manager.

- [ ] Optionally use SPM as the dependency manager (opt in via feature)
- [ ] Add migration for existing apps to use SPM

### PR DESCRIPTION
**Pull request by vashworth**
This PR adds initial support for Swift Package Manager (SPM). Users must opt in. Only compatible with Xcode 15+.

Fixes https://github.com/flutter/flutter/issues/146369.

## Included Features

This PR includes the following features:
* Enabling SPM via config 
`flutter config --enable-swift-package-manager`
* Disabling SPM via config (will disable for all projects) 
`flutter config --no-enable-swift-package-manager`
* Disabling SPM via pubspec.yaml (will disable for the specific project)
```
flutter:
  disable-swift-package-manager: true
```
* Migrating existing apps to add SPM integration if using a Flutter plugin with a Package.swift
  * Generates a Swift Package (named `FlutterGeneratedPluginSwiftPackage`) that handles Flutter SPM-compatible plugin dependencies. Generated package is added to the Xcode project.
* Error parsing of common errors that may occur due to using CocoaPods and Swift Package Manager together
* Tool will print warnings when using all Swift Package plugins and encourage you to remove CocoaPods

This PR also converts `integration_test` and `integration_test_macos` plugins to be both Swift Packages and CocoaPod Pods.

## How it Works
The Flutter CLI will generate a Swift Package called `FlutterGeneratedPluginSwiftPackage`, which will have local dependencies on all Swift Package compatible Flutter plugins.  

The `FlutterGeneratedPluginSwiftPackage` package will be added to the Xcode project via altering of the `project.pbxproj`. 

In addition, a "Pre-action" script will be added via altering of the `Runner.xcscheme`. This script will invoke the flutter tool to copy the Flutter/FlutterMacOS framework to the `BUILT_PRODUCTS_DIR` directory before the build starts. This is needed because plugins need to be linked to the Flutter framework and fortunately Swift Package Manager automatically uses `BUILT_PRODUCTS_DIR` as a framework search path.

CocoaPods will continue to run and be used to support non-Swift Package compatible Flutter plugins.

## Not Included Features

It does not include the following (will be added in future PRs):
* Create plugin template
* Create app template
* Add-to-App integration

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Comment by stuartmorgan** [comment]
> * Disabling SPM via pubspec.yaml (will disable for the specific project)

What's the use case for project-level control? It would be nice if we could minimize the number of knobs and switches that we have to support.

**Comment by vashworth** [comment]
> > * Disabling SPM via pubspec.yaml (will disable for the specific project)
> 
> What's the use case for project-level control? It would be nice if we could minimize the number of knobs and switches that we have to support.

By default, it'll use CocoaPods and Swift Package Manager together, but they don't actually work together. For example, if a SPM-supported plugin and CocoaPod-only plugin is being used and both plugins have a same transitive dependency, it can cause errors about duplicate symbols or redefined modules. The only workaround then is to remove one of the plugins or fallback to just using CocoaPods only (aka via disabling SPM in the pubspec.yaml). Hypothetically could disable it globally, but that would be annoying to have to run the enable/disable command per project.

**Comment by stuartmorgan** [comment]
Ah right, I forgot about that limitation. Makes sense!

File: packages/flutter_tools/bin/podhelper.rb
Comment (stuartmorgan) [inline packages/flutter_tools/bin/podhelper.rb:326]: Do you know off-hand what counts as "when plugins change" [here](https://github.com/flutter/flutter/blob/477ebd831c1253d2e19ac8084963f3088282ddd3/packages/flutter_tools/lib/src/macos/cocoapods.dart#L309-L314)? I'm wondering if we re-run `pod install` if the version of an existing plugin changes, or only when the set of plugins change. If the latter, it seems like we could have issues when an existing project's dependencies add SPM support.

(This may well be answered elsewhere in the PR, I'm just doing some initial skimming and asking questions as they come up.)

Comment (vashworth) [reply packages/flutter_tools/bin/podhelper.rb:326]: We use `Fingerprinter` that checks various files. I added the generated Swift Package manifest as a path so that if it changes, `pod install` will also happen

https://github.com/flutter/flutter/blob/8c04fd74c8278ad400951098fc963e8bfd6282f0/packages/flutter_tools/lib/src/macos/cocoapod_utils.dart#L32-L49

**Comment by stuartmorgan** [comment]
I converted `url_launcher_ios` following your excellent instructions (one minor issue, but that was entirely due to me missing part of a step), and it worked in a test project flawlessly. Manually verified end-to-end plugin functionality, and no `Pods` created!

**Review by stuartmorgan-g — COMMENTED** [review]
I haven't reviewed the Xcode project editor or the tests yet, but posting comments for the rest now. Overall this looks really good!

File: packages/flutter_tools/lib/src/migrations/swift_package_manager_integration_migration.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/migrations/swift_package_manager_integration_migration.dart:166]:
```suggestion
          throwToolExit('Settings were not updated correctly.');
```

File: packages/flutter_tools/lib/src/migrations/swift_package_manager_integration_migration.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/migrations/swift_package_manager_integration_migration.dart:224]:
```suggestion
      throwToolExit('Unable to get Xcode project info.');
```

File: packages/flutter_tools/lib/src/migrations/swift_package_manager_integration_migration.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/migrations/swift_package_manager_integration_migration.dart:119]:
```suggestion
        throwToolExit('Xcode project not found.');
```

**Review by jmagman — COMMENTED** [review]
> Create app template

I know this PR is already huge, but can we add the app template change into this as well?  It would be good for newly `flutter create`d apps to not need to go through the migration path.

Also, might be good to check in at least one example/integration test app so we can see what this looks like migrated (and then follow up in another PR with all the projects migrated)

**Comment by jmagman** [comment]
@loic-sharma fyi if you're interested in the first draft of this.

**Comment by vashworth** [comment]
@stuartmorgan I had to make a couple updates to the instructions.

1. Library names in the Package.swift manifest can't contain underscores, so I changed them to be hyphen separated. 

    Reason:  If the package is linked dynamically, SPM automatically uses the library name as the `CFBundleIdentifier`.
    > The bundle ID string must contain only alphanumeric characters (A–Z, a–z, and 0–9), hyphens (-), and periods (.)
    
    https://developer.apple.com/documentation/bundleresources/information_property_list/cfbundleidentifier
    
    Apparently, it will cause error when you try to upload to app store: https://stackoverflow.com/questions/62477747/swift-package-manager-problem-with-bundle-ids

2. I decided to change the way we do units tests back to using modulemaps. Using a package's testTarget directly doesn't launch a simulator and install the app, and some of our tests assume that it does. 

    Also it seems some tests may require permissions from Info.plist, entitlements, etc which is again part of the app. 
    
    Also, this keeps any test-specific dependencies (like OCMock) out of the plugin itself - which will improve speeds to users consuming the plugin since it won’t download extra unneeded dependencies.
 

I updated the instructions to reflect these new changes, could you try it again when you have time?
https://github.com/flutter/flutter/pull/146256#issuecomment-2038565077

File: packages/flutter_tools/lib/src/xcode_project.dart
Comment (jmagman) [reply packages/flutter_tools/lib/src/xcode_project.dart:124]: I think she's adding it to the .gitignore.  Same comment for iOS, instead can we add it to an ephemeral directory that is already gitignored, and that `flutter clean` already knows about?

File: packages/flutter_tools/bin/xcode_backend.dart
Comment (vashworth) [reply packages/flutter_tools/bin/xcode_backend.dart:383]: It does not show unless verbose, but it's not really necessary or anything so I'll just remove it.

File: packages/flutter_tools/lib/src/ios/mac.dart
Comment (vashworth) [reply packages/flutter_tools/lib/src/ios/mac.dart:172]: Sure, I initially decided against it because it doesn't use any of the functionality of `ProjectMigrator` and `SwiftPackageManagerIntegrationMigration` needs to be async, but your reasoning makes sense. I'll probably make a new PR that converts `ProjectMigrator` to be async

File: packages/flutter_tools/lib/src/macos/cocoapod_utils.dart
Comment (vashworth) [reply packages/flutter_tools/lib/src/macos/cocoapod_utils.dart:37]: It doesn't error if the file is missing but does print out a trace message about not being able to find it, which is why I added it here

https://github.com/flutter/flutter/blob/137cb4be670d48f3d9b45a5a6c6cd50df488b4a7/packages/flutter_tools/lib/src/base/fingerprint.dart#L68

File: packages/flutter_tools/lib/src/macos/swift_package_manager.dart
Comment (vashworth) [reply packages/flutter_tools/lib/src/macos/swift_package_manager.dart:147]: Not yet, blocked by https://github.com/flutter/flutter/issues/146204. I'll add a TODO

File: packages/flutter_tools/lib/src/macos/swift_packages.dart
Comment (vashworth) [reply packages/flutter_tools/lib/src/macos/swift_packages.dart:37]: No this file will not really be editable by the user since it's generated. Unfortunately, swift packages don't have as much control as CocoaPods did. 

So `use_frameworks!` would tell CocoaPods to use dynamic libraries rather than static - that's not possible with SPM. You can [define whether the package itself](https://developer.apple.com/documentation/packagedescription/product/library(name:type:targets:)#parameters) is static or dynamic, but not dependencies.

Also, FYI, SPM recommends not setting what type and letting Xcode decide. It seems to default to static. 

I did confirm that privacy manifest are still included in the app even if statically linked. SPM puts them in a `.bundle` in the Runner.app directory. I also tested that Xcode's Generate Report archival process found them.

File: packages/flutter_tools/lib/src/migrations/swift_package_manager_integration_migration.dart
Comment (vashworth) [reply packages/flutter_tools/lib/src/migrations/swift_package_manager_integration_migration.dart:119]: Does it matter if it's `throwToolExit` since this is in a try catch where the catch will rethrow as a `throwToolExit`?

Comment (jmagman) [reply packages/flutter_tools/lib/src/migrations/swift_package_manager_integration_migration.dart:119]: Ah, missed that.  I think what you have is fine then?

Comment (vashworth) [reply packages/flutter_tools/lib/src/migrations/swift_package_manager_integration_migration.dart:119]: Yeah, it's fine. I just remembered my unit tests expect `throwsToolExit` so it confirms it's throwing the correct thing

File: packages/flutter_tools/lib/src/xcode_project.dart
Comment (vashworth) [reply packages/flutter_tools/lib/src/xcode_project.dart:124]: Oh good idea, I totally forgot about the ephemeral directory

**Comment by stuartmorgan** [comment]
@vashworth I migrated `file_selector_ios` and `file_selector_macos` with the new instructions, and everything worked!

(I kept the weird structure where the headers that are supposed to be test-only are actually public; I don't remember why exactly we have to do that for our modulemap-based plugins, but it's a pre-existing problem, and longer term Swift migrations will eliminate the issue so I didn't worry about it.)

File: packages/flutter_tools/lib/src/ios/mac.dart
Comment (vashworth) [reply packages/flutter_tools/lib/src/ios/mac.dart:907]: I couldn't find very good documentation on Podfile.lock - https://stackoverflow.com/a/27955017 was the best I could find. Let me know if you want me to adjust it more

File: packages/flutter_tools/lib/src/macos/cocoapod_utils.dart
Comment (vashworth) [inline packages/flutter_tools/lib/src/macos/cocoapod_utils.dart:36]: I change this if statement to try and make it easier to read, it kept confusing me lol. I also added unit tests for this file/function to make sure it remained the same. I guess it was never unit tested directly before...
https://github.com/flutter/flutter/blob/503ae5b9324028d3c4bad797873a9d13469ab33b/packages/flutter_tools/test/general.shard/macos/cocoapod_utils_test.dart

**Review by jmagman — APPROVED** [review]
This looks great!  It's behind a feature flag, so I think the best thing would be to merge it and iterate from there based on what we discover from real-world usage.

**Comment by stuartmorgan** [comment]
+1 to that! There's good test coverage so we're well positioned to iterate, all of my manual tests went smoothly, and we probably won't find weird edge-case bugs until we actually have some real-worth early adopters trying it.

Once it lands we can migrate our own plugins ASAP (I have a couple done locally, and should have time this week to do more), to accelerate real-world trials.

**Comment by vashworth** [comment]
> Once it lands we can migrate our own plugins ASAP (I have a couple done locally, and should have time this week to do more), to accelerate real-world trials.

@stuartmorgan I created https://github.com/flutter/flutter/issues/146922 to track migrating our plugins. I assigned you the ones you mentioned you've already worked on. I assigned myself ones I've already worked on. Once this lands, feel free to migrate any of them - just assign yourself so we don't duplicate work

**Comment by stuartmorgan** [comment]
Thanks, I was going to make that bugtree myself but you beat me to it :)

**Pull request merged**
